File: src/wf.nim
import std/os
import std/parseopt
import std/strutils
import db_connector/db_sqlite
import std/times
import std/options
import noise

let appName = extractFileName(getAppFileName())

proc opendb(readonly: bool = false): DbConn =

  var 
    appdir: string = getAppDir()
    dbpath: string

  if "projects" in appdir:

    dbpath = parentDir(appdir) / ".local/share/medical.db" 

  else:

    dbpath = getHomeDir() / ".local/share/medical.db"
    
  let db = open(dbpath, "", "", "")

  if readonly:

    db.exec(sql"PRAGMA query_only = true")

  else:

    db.exec(sql"pragma foreign_keys = on")
    
  result = db

proc help(subcommand: string, exitCode: int = QuitSuccess, ) =

  case subcommand

    of "select":

      echo "Usage: " & appName & " select [OPTIONS] [SEARCH]\n"

      echo """
  Print records from the weight table.  Optionally limit what records are printed by including only records that satisfy the search
  criteria specified in the SEARCH parameter and have dates falling between the BEGIN_DATE and END_DATE.

  The SEARCH parameter may be blank or contain the operators *, (), AND, OR and NOT.  Valid date format is yyyy-MM-dd.  BEGIN_DATE defaults to 52 weeks ago.
  END_DATE defaults to the current date.

  Options:
    -b[=|:]BEGIN_DATE
    -e[=|:]END_DATE
    -h, Show this message and exit.

  Single letter options that take a parameter require an equal sign or a colon between the option and the argument.
  """

    of "insert":

      echo "Usage: " & appName & " insert [OPTIONS] [FOOD]\n"
      echo """
  Insert records into the weight table.  The defaults are DATE = today, WEIGHT = zero and [FOOD] = "". 

  Options:
    -d[=|:]DATE
    -w[=|:]WEIGHT
    -h Show this message and exit.

  Single letter options that take a parameter require an equal sign or a colon between the option and the argument.
  """

    of "update":

      echo "Usage: " & appName & " update [OPTIONS] ID.\n"
      echo """
  Update the weight record having the given ID. The ID must be an integer greater than or equal to zero.

  Options:
    -h  Show this message and exit.
  """

    of "delete":

      echo "Usage: " & appName & " delete [OPTIONS] ID\n"
      echo """
  Delete the weight record having the given ID.  The ID must be an integer greater than or equal to zero.

  Options:
    -h Show this message and exit.
  """

    of "main":

      echo "Usage: " & appName & " [OPTIONS] COMMAND [ARGS]...\n"
      echo """
  Maintain a weight and food database.

  Options:
    -h  Show this message and exit.

  Commands:
    select  Print records from the weight table.
    insert  Insert a record into the weight table.
    update  Update a weight record having a given ID.
    delete  Delete a weight record having a given ID.

    For help on any command follow the command with -h.
  """

  quit(exitCode)

proc select(begintime, endtime: DateTime, searchstr: Option[string]) =

  var sqlstring: SqlQuery

  var db = opendb(readonly = true)

  var rows: seq[Row]

  if searchstr.isNone():

    sqlstring = sql"select id, date(date) as date, weight, food from weight where date >= ? and date <= ? order by date"

    rows = db.getAllRows(sqlstring, begintime.format("yyyy-MM-dd"), endtime.format("yyyy-MM-dd"))

  elif searchstr.get == "":
  
    sqlstring = sql"select id, date(date) as date, weight, food from weight where food = '' and date >= ? and date <= ? order by date"

    rows = db.getAllRows(sqlstring, begintime, endtime)
    # rows = db.getAllRows(sqlstring, begintime.format("yyyy-MM-dd"), endtime.format("yyyy-MM-dd"))
  
  elif searchstr.isSome():

    try:    

      var searchstring = searchstr.get

      if searchstring.startsWith("NOT "):

        searchstring.removePrefix("NOT ")

        sqlstring = sql"""select id, date(date) as date, weight, food from weight
                where id not in ( select docid from weightft where weightft match ? )
                and date >= ? and date <= ? order by date"""
      else:

        sqlstring = sql"""select id, date(date) as date, weight, food from weight
                where id in ( select docid from weightft where weightft match ? )
                and date >= ? and date <= ? order by date"""

      rows = db.getAllRows(sqlstring, searchstring, begintime.format("yyyy-MM-dd"), endtime.format("yyyy-MM-dd"))

    except DbError as e:

     if "fts5" in e.msg:
        
       quit(appName & " select: Encountered a search error. Use only the operators *, (), AND, OR and NOT in the search parameter.  The valid date format is yyyy-MM-dd.")
      
  for row in rows:

    echo alignleft(row[0],6), alignleft(row[1], 11),  align(row[2], 7), "  ", row[3]
  
proc insert(date: string, weight: float, food: string = "") =

  var db = opendb()

  var sql = sql"insert into weight (date, weight, food) values(?, ?, ?)"

  try:

    db.exec(sql, date, weight, food)

  except DbError as e:

    if e.msg.contains("UNIQUE"):

      quit(appName & ": There is already an entry for that date: " & date)
  
  finally:

    sql = sql"select last_insert_rowid()"

    let id = db.getValue(sql)

    echo "Successfully inserted a row with the id " & id    

    db.close()

proc update(id: Natural) =

  var sql = sql"select date, weight, food from weight where id = ?"

  let db = opendb()

  var row: Row

  try:

    row = db.getRow(sql,id) 

    doAssert row != @["", "", ""]

  except AssertionDefect:

    quit("$1 update: No record with the ID '$2' was found. Are you sure it exists?" % [appName, $id]) 

  var noise = Noise.init()

  noise.setPrompt("Date: ")

  var date = row[0]

  noise.preloadBuffer(date)

  if noise.readLine():

    date = noise.getLine()

    try:
    
      date = date.parse("yyyy-MM-dd").format("yyyy-MM-dd")

    except TimeParseError as e:

      quit(appName & " update: " & e.msg)

  noise.setPrompt("Weight: ")

  var weight = row[1]

  noise.preloadBuffer(weight)

  if noise.readLine():

    weight = noise.getLine()

    try:

      let f = parseFloat(weight)

      doAssert f < 300 and f >= 0

    except ValueError, AssertionDefect:

      quit(appName &  " update: The weight must be a floating point number less then 300 and greater then or equal to 0.  The decimal point is optional.")
   
  noise.setPrompt("Food: ")

  var food = row[2]

  noise.preloadBuffer(food)

  if noise.readLine():

    food = noise.getLine()

  sql = sql"update weight set date = date(?), weight = ?, food = ? where id = ?"

  if db.tryExec(sql, date, weight, food, id):

    echo "Update of the weight record with ID: ", id, " succeeded."

  else:

    quit("Update of the weight record with ID: " & $id & " failed.")

proc delete(id: int) =

  let db = opendb()

  try:
  
    discard db.getValue(sql"select id from weight where id = ?", id).parseInt()

  except ValueError:

    quit("The weight table does not have a record with the ID: " & $id & ".")

  db.exec(sql"BEGIN")

  if db.tryExec(sql"delete from weight where id = ?", id):

    stdout.write("Confirm the deletion of the weight record with the ID: " & $id & " by pressing y. ")
    
    var answer = getch()

    if answer == 'y':

      db.exec(sql"COMMIT")

      echo "\nDeleted the record with the ID: ", id, "."

    else:

      db.exec(sql"ROLLBACK")

      echo "\nDid not delete the record with the ID: ", id, "."

  db.close()

proc main() =

  if paramCount() == 0:

    help("main")

  case paramStr(1)

    of "select":

      if "-h" in commandLineParams():

        help("select")

      var
        endtime: DateTime = now() # + initDuration(days = 1)
        begintime: DateTime = endtime - initDuration(weeks = 52)
        search: Option[string]

      var p = initOptParser()

      while true:

        p.next()

        case p.kind

          of cmdEnd: break

          of cmdLongOption:

            quit(appName & ": " & "'--" & p.key & "' Long options are not supported.")

          of cmdShortOption:

            case p.key

              of "b":

                try:

                  begintime = parse(p.val, "yyyy-MM-dd")

                except TimeParseError as e:

                  quit(appName & ": " & e.msg)

              of "e":

                try:

                  endtime = parse(p.val, "yyyy-MM-dd")

                except TimeParseError as e:

                  quit(appName & ": " & e.msg)

          of cmdArgument:

            if p.key != "select":

              begintime = endtime - initDuration(weeks = 5200)

              if search.isNone():

                search = some(p.key)

              else:

                search = some(search.get & " " & p.key)
      
      if endtime < begintime:

        quit("$1 select: The END_DATE must be later than the BEGIN_DATE." % appName, QuitFailure)

      select(begintime, endtime, search)

    of "insert":

      if "-h" in commandLineParams():

        help("insert")

      var
        date: string = getDateStr()
        weight: float
        food: string

      var p = initOptParser()

      while true:

        p.next

        case p.kind

          of cmdEnd: break

          of cmdLongOption:

              quit(appName & ": " & "'--" & p.key & "' Long options are not supported.")

          of cmdShortOption:

            case p.key

              of "d":

                try:

                  if p.val.parse("yyyy-MM-dd") > now():

                    quit("$1 insert: DATE '$2' must be in the past." % [appName, p.val])

                except ValueError as e:

                  quit("$1 insert: Invalid date. $2." % [appName, e.msg])

                date = p.val

              of "w":

                try:

                  weight = p.val.parseFloat()

                except ValueError:

                  quit("$1 insert: Invalid weight '$2'.  WEIGHT must be a floating point number." % [appName, p.val])
            
          of cmdArgument:

            if p.key != "insert":

              food.add(" " & p.key)

      food = food.strip()  

      insert(date, weight, food)

    of "update":

      if "-h" in commandLineParams():

        help("update")

      var id : Natural

      try:

        id = parseInt(paramStr(2))

      except:
        
        quit("$1 insert: ID '$2' must be an integer." % [appName, paramStr(2)])

      update(id)

    of "delete":

      if "-h" in commandLineParams():

        help("delete")

      var id: Natural

      try:

        id = parseInt(paramStr(2))

      except:

        quit("$1 delete: ID '$2' must be an integer." % [appName, paramStr(2)])

      delete(id)

    else:

      echo "$1 Unknown command: $2" % [appName, paramStr(1)], "\n"

      help("main", QuitFailure)

when isMainModule:
  
  main()

File: src/wf.py
#!/usr/bin/python

# vi:syntax=python

import sqlite3
import click
from tabulate import tabulate
from pathlib import Path
import datetime as dt
from prompt_toolkit import prompt
from prompt_toolkit.validation import Validator, ValidationError
import re

def dbopen(readonly=False):

    try:

        programdir = Path(__file__).parent.resolve()
        
        if 'projects' in str(programdir):

            dbfile = Path().home() / 'projects' / '.local' / 'share' / 'medical.db'
        
        else:

            dbfile = Path().home() / '.local' / 'share' / 'medical.db'

        uri = 'file:' + str(dbfile)

        if readonly:

            db = sqlite3.connect(uri + '?mode=ro', uri=True)

        else:

            db = sqlite3.connect(uri, uri=True)

        db.execute("pragma foreign_keys = on")

        db.row_factory = sqlite3.Row
            
    except sqlite3.OperationalError as e:

        click.echo(f'{str(e).capitalize()}: {dbfile}')

        exit(1)

    return db

@click.group()
def weight():

    '''Maintain a weight and food database.
    '''


@weight.command()
@click.argument('search', required=False, type=click.STRING(None))
@click.option('--begin_date', '-b', type=click.DateTime(['%Y-%m-%d']), metavar='[BEGIN_DATE]')
@click.option('--end_date', '-e', type=click.DateTime(['%Y-%m-%d']), metavar='[END_DATE]')
def select(search, begin_date, end_date):

    """Print records from the weight table.  Optionally limit what records are printed by including only records that satisfy the search
    criteria specified in the SEARCH parameter and have dates falling between the BEGIN_DATE and END_DATE.
    
    The SEARCH parameter may be blank or contain the operators *, (), AND, OR and NOT.  Valid date format is %Y-%m-%d.  START_DATE defaults to 365 days ago.
    END_DATE defaults to the current date.
    """

    db = dbopen(True)

    if begin_date is None and end_date is not None:

        begin_date = end_date - dt.timedelta(days=365)

    if begin_date is None and end_date is None:

        end_date = dt.datetime.now()
        begin_date = end_date - dt.timedelta(days=365)

    if end_date is None:

        end_date = dt.datetime.now() #+ dt.timedelta(days=1)

    #include begin and end dates.  Don't know why the begin date is not included unless I subtract one from it.

    begin_date = begin_date - dt.timedelta(days=1)

#    end_date = end_date + dt.timedelta(days=1)
    
    if search is None:

        sql = '''select id, date(date) as date, weight, food from weight 
                where date >= ? and date <= ? order by date'''

        parameters = (begin_date, end_date)

    elif search == "":
    
        sql = '''select id, date(date) as date, weight, food from weight 
                where food = "" and date >= ? and date <= ? order by date'''

        parameters = (begin_date, end_date)

    else:

        sql = '''select id, date(date) as date, weight, food from weight 
                where id in ( select docid from weightft where weightft match ? )
                and date >= ? and date <= ? order by date'''

        parameters = (search, begin_date, end_date)

    try:

        table = db.execute(sql, parameters).fetchall()

        if len(table) > 0:

            table = db.execute(sql, parameters)

        else:

            click.echo(f"None found.")

            exit(0)
    
    except sqlite3.OperationalError:

        click.echo('Search error: Use only an empty string or the operators *, (), AND, OR and NOT in the search parameter.  The valid date format is %Y-%m-%d.')

        exit(1)

    names = [description[0].upper() for description in table.description]

    print(tabulate(table,names))

#    print(sql,parameters)

@weight.command()
@click.option('--date', '-d', type=click.DateTime(['%Y-%m-%d']), default=str(dt.date.today()))
@click.option('--weight', '-w', type=click.FLOAT, default=0)
@click.option('--food', '-f', default='')
def insert(date, weight, food):

    '''Insert a record into the weight database.
    '''

    date = date.strftime('%Y-%m-%d')

    sql = '''insert into weight (date, weight, food) values(?, ?, ?)'''

    db = dbopen()

    try:

        with db:

            cursor = db.cursor()

            cursor.execute(sql, (date, weight, food))

            click.echo(f"Insertion of weight {cursor.lastrowid} succeeded.")

    except sqlite3.Error as e:

        if 'FOREIGN' in e.args[0]:

            click.echo(f"Medication '{name}' does not exist.")

        else:

            click.echo(f'Weight insertion {e}')
        
        exit(1)

class UpdateDateValidator(Validator):

    def validate(self, document):

        text = document.text
        
        isValidDate = True

        try:

            dt.datetime.strptime(text, '%Y-%m-%d')

        except ValueError:

            isValidDate = False
        
        if not re.match(r'\d{4}-\d{2}-\d{2}', text):

            isValidDate = False

        if not isValidDate:

            raise ValidationError(message = 'Time must be of the form: %Y-%m-%d')

class UpdateFloatValidator(Validator):

    def validate(self, document):

        text = document.text

        try:

            float(text)

        except ValueError:

            raise ValidationError(message = 'Must be a floating point number')

@weight.command()
@click.argument('id', type=int)
def update(id):

    '''Update the weight record having the given ID.
    '''

    db = dbopen()

    sql = '''select strftime('%Y-%m-%d',date), weight, food from weight where id = ?'''

    weight = db.execute(sql, (id,)).fetchone()
    
    if not weight:

        click.echo(f"Weight {id} not found.")

        exit(1)
        
    ( date, weight, food ) = weight

    if food == None:

        food = ''

    sql = "update dose set name = ?, datetime = datetime(?), comment = ? where id = ?"
    
    click.echo(f"Updating weight record: {id}")

    date = prompt("Date: ", default=date, validator=UpdateDateValidator(), validate_while_typing=False)

    weight = prompt("Weight: ", default=str(weight), validator=UpdateFloatValidator(), validate_while_typing=False)

    food = prompt("Food: ", default=food)
    
    sql = "update weight set date = date(?), weight = ?, food = ? where id = ?"

    try:

        with db:

            db.execute(sql, (date, weight, food, id))
                
    except sqlite3.Error as e:

        click.echo(f'Update weight: {e}')
    
        exit(1)

    click.echo(f'Update of weight {id} succeeded.')


@weight.command()
@click.argument('id', type=int)
def delete(id):

    '''Delete the weight record having the given ID.  The ID must be an integer.
    '''

    db = dbopen()

    sql = 'delete from weight where id = ?'

    try:

        with db:

            deleted = db.execute(sql,(id,))

            if deleted.rowcount == 0:

                click.echo(f"Deletion of weight {id} failed. Are you sure it exists?")

                exit(1)

            else:

                if click.confirm(f'Confirm deletion of weight record with ID number: {id}', default=False, show_default=True):
                
                    click.echo(f'Deletion of weight {id} succeeded.')

                else:

                    click.echo(f'Confirmation failed. weight {id} not deleted.' )

                    db.rollback()

    except sqlite3.Error as e:

        click.echo(e)

        exit(1)

if __name__ == '__main__':
        weight()
